### PR TITLE
Async

### PR DESCRIPTION
**Pull request by abadams**
Adds the ability to schedule Funcs to be computed asynchronously. From Func::async docs:

    /** Produce this Func asynchronously in a separate
     * thread. Consumers will be run by the task system when the
     * production is complete. If this Func's store level is different
     * to its compute level, consumers will be run concurrently,
     * blocking as necessary to prevent reading ahead of what the
     * producer has computed. If storage is folded, then the producer
     * will additionally not be permitted to run too far ahead of the
     * consumer, to avoid clobbering data that has not yet been
     * used.
     *
     * Take special care when combining this with custom thread pool
     * implementations, as avoiding deadlock with producer-consumer
     * parallelism requires a much more sophisticated parallel runtime
     * than with data parallelism alone. It is strongly recommended
     * you just use Halide's default thread pool, which guarantees no
     * deadlock and a bound on the number of threads launched.
    *//

Not ready for review (there are a handful of TODOs), but this is at the point where I want continuous buildbot coverage.

### CODE REVIEW

**Comment by steven-johnson** [comment]
How close are we getting on this?

**Comment by zvookin** [comment]
Should no longer be deadlocking and does not depend on thread local storage. Needs review and buildbot coverage, but should be mostly ready to merge. I'm going to see if we can tighten up the code a little bit as well to make it perform as well as or better than master on performance_inner_loop_parallel. (It is currently a bit slower on that test. Not worrisomely so however.)

File: src/runtime/HalideRuntime.h [hunk 1]
```diff
@@ -130,6 +127,103 @@ extern void halide_shutdown_thread_pool();
...
+extern int halide_semaphore_release(struct halide_semaphore_t *, int n);
+extern bool halide_semaphore_try_acquire(struct halide_semaphore_t *, int n);
+typedef int (*halide_semaphore_init_t)(struct halide_semaphore_t *, int);
+typedef int (*halide_semaphore_release_t)(struct halide_semaphore_t *, int);
+typedef bool (*halide_semaphore_try_acquire_t)(struct halide_semaphore_t *, int);
+
+
+/** A task representing a serial for loop evaluated over some range. */
+typedef int (*halide_loop_task_t)(void *user_context, int min, int extent,
+                                  uint8_t *closure, void *);
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: What's the final void* argument? Shouldn't it be named?

File: src/runtime/HalideRuntime.h [hunk 2]
```diff
@@ -130,6 +127,103 @@ extern void halide_shutdown_thread_pool();
...
+
+/** A parallel task to be passed to halide_do_parallel_tasks. That
+ * tasks may recursively call halide_do_parallel_tasks, and there may
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: s/tasks/task/

File: src/runtime/HalideRuntime.h [hunk 3]
```diff
@@ -130,6 +127,103 @@ extern void halide_shutdown_thread_pool();
...
+ * be complex dependencies between seemingly unrelated tasks expressed
+ * using semaphores. If you are using a custom task system, care must
+ * be taken to avoid potential deadlock. This can be done by carefully
+ * respecting the static metadata at the end of the task struct.*/
+struct halide_parallel_task_t {
+    // The function to call. It takes a user context, a min and extent, and a closure.
+    halide_loop_task_t fn;
```
Comment (steven-johnson) [inline src/runtime/HalideRuntime.h]: The actual fn has 5 args, not 4, as the comment would imply

File: test/correctness/async.cpp [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    }
+    if (f < 0) return 3;
+    return x;
+}
+HalideExtern_1(int, expensive, int);
+
+int main(int argc, char **argv) {
+
+    // Basic compute-root async producer
+    if (1) {
```
Comment (steven-johnson) [inline test/correctness/async.cpp]: I presume the `if (1)` are for testing/debugging convenience, but we should probably remove them before landing

File: test/correctness/async_device_copy.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
+#include "Halide.h"
+#include "halide_benchmark.h"
+
+using namespace Halide;
+
+// Return zero, slowly
+Expr expensive_zero(Expr x, Expr y, Expr t, int n) {
+    // Count how many Fermat's last theorem counterexamples we can find using n trials.
```
Comment (steven-johnson) [inline test/correctness/async_device_copy.cpp]: I have discovered a truly marvelous proof of this, which this column-width is too narrow to contain

File: test/correctness/parallel_fork.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    f(x, y) = halide_externs::five_ms(x + y);
+    g(x, y) = halide_externs::five_ms(x - y);
+
+    both(x, y, z) = select(z == 0, f(x, y), g(x, y));
+
+    both.compute_root().bound(z, 0, 2);
+    switch (schedule) {
+    case Serial:
+        f.compute_root();
```
Comment (steven-johnson) [inline test/correctness/parallel_fork.cpp]: indentation

File: src/AsyncProducers.cpp [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+        }
+    }
+
+public:
+    ForkAsyncProducers(const map<string, Function> &e) : env(e) {}
+};
+
+// Lowers semaphore initialization from a call to
+// "halide_make_semaphore" to an alloca followed by a call into the
+// runtime to initialize. TODO: what if something crashes before
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/AsyncProducers.cpp]: @abadams Do you mind looking over the new comment on InitializeSemaphores re: error handling in async. I added code to propagate errors across all branches of a fork. I'm not 100% sure it handles complex nested async/par_for structures however. (The situation is that some other part of the tree needs a semaphore that is not going to be released because another part failed, but the failure will not have poisoned the part of the tree that is waiting. I'm not convinced this can actually happen. Probably the thing to do is to take the async_parallel test and add errors into it.)

File: src/AsyncProducers.cpp [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+                return ProducerConsumer::make(name, is_producer, body);
+            } else if (first && rest) {
+                return Block::make(make_producer_consumer(name, is_producer, block->first, scope),
+                                   make_producer_consumer(name, is_producer, block->rest, scope));
+            } else if (first) {
+                return Block::make(make_producer_consumer(name, is_producer, block->first, scope), block->rest);
+            } else if (rest) {
+                return Block::make(block->first, make_producer_consumer(name, is_producer, block->rest, scope));
+            } else {
+                // Used on neither side?!
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/AsyncProducers.cpp]: @abadams Any thoughts?

File: src/InjectHostDevBufferCopies.cpp [hunk 9]
```diff
@@ -332,7 +333,11 @@ class InjectBufferCopiesForSingleBuffer : public IRMutator2 {
     // leaf.
 
     Stmt visit(const For *op) override {
-        // All copies happen at the same loop level as the allocation
+        // All copies happen at the same loop level as the allocation (TODO: This prevents hoisting allocations)
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/InjectHostDevBufferCopies.cpp]: @abadams Do you mind weighing in on this?

File: src/StorageFolding.cpp [hunk 10]
```diff
@@ -107,11 +107,13 @@ class FoldStorageOfFunction : public IRMutator {
                 Expr no_wraparound = mins[dim] + extents[dim] <= factor;
 
                 Expr valid_min = old_min;
-                if (!dynamic_footprint.empty()) {
+                if (false && !dynamic_footprint.empty()) {
```
[the remark this answers is not in the document]
Comment (zvookin) [reply src/StorageFolding.cpp]: @abadams Do you mind weighing in on this?

[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/StorageFolding.cpp]: at a minimum do something like
```
// explain why do_this_this is always off
constexpr bool do_this_thing = false;
if (do_this_thing && ...) {
}
```

File: src/AsyncProducers.cpp [hunk 11]
```diff
@@ -14,81 +14,81 @@ using std::string;
...
+            return LetStmt::make(op->name, op->value, body);
         }
     }
 
-    void visit(const For *op) {
+    Stmt visit(const For *op) {
         Stmt body = mutate(op->body);
         if (is_no_op(body)) {
-            stmt = body;
+             return body;
```
Comment (steven-johnson) [inline src/AsyncProducers.cpp]: extra space

File: src/AsyncProducers.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (abadams) [reply src/AsyncProducers.cpp]: No, I don't remember. I guess this path would delete producer consumer nodes that don't actually contain a use. Is such a thing possible? Perhaps with specializations that simplify away things...

File: src/InjectHostDevBufferCopies.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (abadams) [reply src/InjectHostDevBufferCopies.cpp]: I have no memory of this place...

I can see that split store_at compute_at might need copies to happen at the compute_at level instead of the allocation level, but that's just not what the model is. Maybe just revert this todo.

File: src/AsyncProducers.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (abadams) [reply src/AsyncProducers.cpp]: I can't think through the possibilities in my head reliably for this sort of thing... Perhaps this feature needs a deadlock/error-handling fuzzer: construct random combinations of nested parallelism and check they can run, and check they clean up properly when errors are thrown internally. We would have to run it with a watchdog to avoid wedging buildbots though (SIGALRM?)

File: src/StorageFolding.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (abadams) [reply src/StorageFolding.cpp]: It looks like async make this error checking insufficient, but we never figured out how to make it work again, hence the temporary if (false).

File: src/Func.cpp [hunk 12]
```diff
@@ -1546,11 +1546,22 @@ Stage &Stage::reorder(const std::vector<VarOrRVar>& vars) {
     vector<Expr> &args = definition.args();
     vector<Expr> &values = definition.values();
 
-    // Check whether the operator is associative and commutative.
-    const auto &prover_result = prove_associativity(func_name, args, values);
-    const bool is_associative_and_commutative
-            = prover_result.associative() && prover_result.commutative();
+    bool is_associative_and_commutative = false;
 
+    bool any_rvars = false;
```
Comment (jingpu) [inline src/Func.cpp]: I think we can just check whether there are any RVar in the input to this `reorder` function. There is no point to run the prover if a user is not trying to reorder two RVars.

Comment (dsharletg) [reply src/Func.cpp]: +1, going even further, it's only necessary if RVars are being reordered w.r.t. each other. A hack would be to only do this check if >= 2 RVars are being reordered, a more invasive change would be to only do this check when an error is about to be emitted.

File: src/AllocationBoundsInference.cpp [hunk 13]
```diff
@@ -128,8 +128,7 @@ class AllocationInference : public IRMutator2 {
         for (map<string, Function>::const_iterator iter = e.begin();
              iter != e.end(); ++iter) {
             Function f = iter->second;
-            if (f.has_extern_definition() &&
```
Comment (abadams) [inline src/AllocationBoundsInference.cpp]: Why this change?

File: src/Bounds.cpp [hunk 14]
```diff
@@ -1807,6 +1854,7 @@ class BoxesTouched : public IRGraphVisitor {
     }
 
     void visit(const Provide *op) {
+      debug(1) << "Got provide for " << Stmt(op) << ".\n";
```
Comment (abadams) [inline src/Bounds.cpp]: Doesn't seem like this should be at debug(1). Also indentation.

File: src/InjectHostDevBufferCopies.cpp [hunk 15]
```diff
@@ -228,6 +228,7 @@ class InjectBufferCopiesForSingleBuffer : public IRMutator2 {
             touching_device = d;
         }
 
+
```
Comment (abadams) [inline src/InjectHostDevBufferCopies.cpp]: Extra blank line here

File: src/StorageFolding.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (abadams) [reply src/StorageFolding.cpp]: This still needs dealing with

File: src/runtime/thread_pool_common.h [hunk 16]
```diff
@@ -79,109 +159,383 @@ struct work_queue_t {
...
+            job->task.extent = 0;
+            continue; // So loop exit is always in the same place.
+        }
+
+        dump_job_state();
+
+        // Find a job to run, prefering things near the top of the stack.
+        while (job) {
+            print_job(job, "", "Considering job ");
+            // Only schedule tasks with enough free worker threads
```
Comment (abadams) [inline src/runtime/thread_pool_common.h]: Is this comment still a sufficient description? This sort of thing has been changing.

File: src/AllocationBoundsInference.cpp (hunk 13 above)
Comment (zvookin) [reply src/AllocationBoundsInference.cpp]: The original code removed the touch tracking done here for any extern with a proxy expression. There isn't anything anywhere else that gets the info from the proxy. Also I'm assuming this code block doesn't actually depend on the bounds themselves but just that there is a touch. I may have this wrong, but I'm fairly certain I needed to make this change to get storage folding to happen in at least one case when working with the hex-dma2 branch.

File: src/StorageFolding.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (zvookin) [reply src/StorageFolding.cpp]: So we're now omitting an error check and emitting unsafe code? Seems there should be a test to prove the error happens. Hopefully I'll be able to find something if I remove the false and see what goes into that block of code in all the tests.

**Comment by zvookin** [comment]
Finally merged. Leaving branch here for some bug fixes forthcoming.
